### PR TITLE
[vcpkg/meson] fix some details

### PR DESCRIPTION
**Pull request by Neumann-A**
- Add ADD_BIN_TO_PATH to vcpkg_install_meson
- Make x86 on x64_x86 a native build (unless uwp)
- Make native compiler unuseable in cross builds. 
- Bonus: Fix single configuration builds. 

changes extracted from #13100

### CODE REVIEW

File: scripts/buildsystems/meson/none.txt [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# native file used to make the build machine compiler unusable
```
[the remark this answers is not in the document]
Comment (Neumann-A) [reply scripts/buildsystems/meson/none.txt:1]: i'll try it out in #13100

File: scripts/cmake/vcpkg_configure_meson.cmake [hunk 2]
```diff
@@ -249,7 +249,7 @@ function(vcpkg_internal_meson_generate_cross_file _additional_binaries) #https:/
     endforeach()
 
     string(APPEND CROSS "[properties]\n")
-    string(APPEND CROSS "skip_sanity_check = true\n")
+    #string(APPEND CROSS "skip_sanity_check = true\n")
```
[the remark this answers is not in the document]
Comment (Neumann-A) [reply scripts/cmake/vcpkg_configure_meson.cmake:252]: seems no longer be necessary. I'll convert to draft until the other PR has run through with the changes.

File: scripts/buildsystems/meson/none.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (Neumann-A) [reply scripts/buildsystems/meson/none.txt:1]: meson: error: unrecognized arguments: --cross-only

[the remark this answers is not in the document]
Comment (Neumann-A) [reply scripts/buildsystems/meson/none.txt:1]: I think that argument is only passed to run_cross_test.py and not meson itself

**Comment by JackBoosY** [comment]
@ras0219 @ras0219-msft Ping for reivew again.

**Comment by venabled** [comment]
@Neumann-A playing around with this PR as I'm trying to get the glib port working on Android.

It looks like some of the options in the internal meson cross file are getting generated with some spaces inbetween some of the dashes, e.g.:

```
c_args = ['-g', '-DANDROID', '-fdata-sections', '-ffunction-sections', '-funwind-tables', '-fstack-protector-strong', '-no-canonical-prefixes', '-D_FORTIFY_SOURCE=2', '-march=armv7-a', '-mthumb', '-Wformat', '-Werror=format-security', '-fPIC', '-', '-target=armv7-none-linux-androideabi21', '-', '-gcc-toolchain=/home/venabled/android-ndk-r21d/toolchains/llvm/prebuilt/linux-x86_64', '-', '-sysroot=/home/venabled/android-ndk-r21d/toolchains/llvm/prebuilt/linux-x86_64/sysroot', '-O0', '-fno-limit-debug-info', '-I"/data/vcpkg/installed/arm-android/include"']
```

Leading to this error during configure:

```
../src/glib-2-4069d62e51.clean/meson.build:1:0: ERROR: Unable to detect linker for compiler "/home/venabled/android-ndk-r21d/toolchains/llvm/prebuilt/linux-x86_64/bin/clang -Wl,--version -g -DANDROID -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -D_FORTIFY_SOURCE=2 -march=armv7-a -mthumb -Wformat -Werror=format-security -fPIC - -target=armv7-none-linux-androideabi21 - -gcc-toolchain=/home/venabled/android-ndk-r21d/toolchains/llvm/prebuilt/linux-x86_64 - -sysroot=/home/venabled/android-ndk-r21d/toolchains/llvm/prebuilt/linux-x86_64/sysroot -O0 -fno-limit-debug-info -I"/data/vcpkg/installed/arm-android/include""
stdout: 
stderr: clang: error: unknown argument '-target=armv7-none-linux-androideabi21'; did you mean '--target=armv7-none-linux-androideabi21'?
clang: error: unknown argument '-gcc-toolchain=/home/venabled/android-ndk-r21d/toolchains/llvm/prebuilt/linux-x86_64'; did you mean '--gcc-toolchain=/home/venabled/android-ndk-r21d/toolchains/llvm/prebuilt/linux-x86_64'?
clang: error: unknown argument '-sysroot=/home/venabled/android-ndk-r21d/toolchains/llvm/prebuilt/linux-x86_64/sysroot'; did you mean '--sysroot=/home/venabled/android-ndk-r21d/toolchains/llvm/prebuilt/linux-x86_64/sysroot'?
clang: error: -E or -x required when input is from standard input
clang: error: -E or -x required when input is from standard input
clang: error: -E or -x required when input is from standard input

```

File: scripts/cmake/vcpkg_configure_meson.cmake [hunk 3]
```diff
@@ -101,15 +101,15 @@ function(vcpkg_internal_meson_generate_native_file_config _config) #https://meso
     set(LIBPATH_${_config} "'${L_FLAG}${CURRENT_INSTALLED_DIR}${PATH_SUFFIX_${_config}}/lib'")
     
     set(NATIVE_${_config} "[properties]\n") #https://mesonbuild.com/Builtin-options.html
-    string(REGEX REPLACE "( |^)(-|/)" ";\\2" MESON_CFLAGS_${_config} "${VCPKG_DETECTED_CMAKE_C_FLAGS_${_config}}")
+    string(REGEX REPLACE "( +|^)(-|/)" ";\\2" MESON_CFLAGS_${_config} "${VCPKG_DETECTED_CMAKE_C_FLAGS_${_config}}")
```
Comment (ras0219) [inline scripts/cmake/vcpkg_configure_meson.cmake:104]:
```suggestion
    string(REGEX REPLACE [=[( +|^)([^"]*("([^\"]|\\|\")*")?)]=] ";\\2" MESON_CFLAGS_${_config} "${VCPKG_DETECTED_CMAKE_C_FLAGS_${_config}}")
```
I think this will match quoted strings with backslash escapes, including forms like `-I"hello world"`, but more testing is required.

Comment (venabled) [reply scripts/cmake/vcpkg_configure_meson.cmake:104]: Will try it out here. Sorry my regex game is pretty weak.

Comment (venabled) [reply scripts/cmake/vcpkg_configure_meson.cmake:104]: Making that change threw this:

```
[DEBUG] CMake Error at scripts/cmake/vcpkg_configure_meson.cmake:104:
[DEBUG]   Parse error.  Expected a newline, got identifier with text "list".
[DEBUG] Call Stack (most recent call first):
[DEBUG]   scripts/ports.cmake:95 (include)
[DEBUG] 
[DEBUG] 
[DEBUG] cmd_execute_and_stream_data() returned 256 after     5939 us
[DEBUG] 
[DEBUG] Compiler information tracking can be disabled by passing --feature-flags=-compilertracking
```

Comment (venabled) [reply scripts/cmake/vcpkg_configure_meson.cmake:104]: Adding in a `+` quantifier seemed to catch the flags with leading `--`

```suggestion
    string(REGEX REPLACE "( +|^)(-+|/)" ";\\2" MESON_CFLAGS_${_config} "${VCPKG_DETECTED_CMAKE_C_FLAGS_${_config}}")
```

Comment (Neumann-A) [reply scripts/cmake/vcpkg_configure_meson.cmake:104]: the error is in get-cmake-vars not here

Comment (Neumann-A) [reply scripts/cmake/vcpkg_configure_meson.cmake:104]: seems to be that the CMake docs are off here a bit. Seems like ^ also matches the direct beginning of the rest of the input after a match.

Comment (Neumann-A) [reply scripts/cmake/vcpkg_configure_meson.cmake:104]: I think this is a strange case where ^ means start of string and start after a matched string. That actually explains a lot.... and CMakes docs seems to be off here which says ^ matches only start of input.
So `string(REGEX REPLACE "( |^)(-)" ";\\1" somevar " ----")` actually results in `;-;-;-;-` instead of the expected `;----`

Comment (Neumann-A) [reply scripts/cmake/vcpkg_configure_meson.cmake:104]: https://gitlab.kitware.com/cmake/cmake/-/issues/16899

Comment (Neumann-A) [reply scripts/cmake/vcpkg_configure_meson.cmake:104]: @ras0219 Your regex can match an empty string due to the combination of  `^` `*` and `?`

**Comment by JackBoosY** [comment]
Ping @Neumann-A @ras0219 for review again.

File: scripts/cmake/vcpkg_configure_meson.cmake [hunk 4]
```diff
@@ -291,14 +292,14 @@ function(vcpkg_internal_meson_generate_cross_file_config _config) #https://meson
 
 
     set(NATIVE_${_config} "[properties]\n") #https://mesonbuild.com/Builtin-options.html
-    string(REGEX REPLACE "( |^)(-|/)" ";\\2" MESON_CFLAGS_${_config} "${VCPKG_DETECTED_CMAKE_C_FLAGS_${_config}}")
+    string(REGEX REPLACE "( |^)(--?|/)" ";\\2" MESON_CFLAGS_${_config} "${VCPKG_DETECTED_CMAKE_C_FLAGS_${_config}}")
```
Comment (Neumann-A) [inline scripts/cmake/vcpkg_configure_meson.cmake:295]:
```suggestion
    string(REGEX REPLACE "( +|^)(--?|/)" ";\\2" MESON_CFLAGS_${_config} "${VCPKG_DETECTED_CMAKE_C_FLAGS_${_config}}")
```

File: scripts/cmake/vcpkg_configure_meson.cmake [hunk 5]
```diff
@@ -291,14 +292,14 @@ function(vcpkg_internal_meson_generate_cross_file_config _config) #https://meson
...
     list(TRANSFORM MESON_CFLAGS_${_config} APPEND "'")
     list(TRANSFORM MESON_CFLAGS_${_config} PREPEND "'")
     list(APPEND MESON_CFLAGS_${_config} "'-I\"${CURRENT_INSTALLED_DIR}/include\"'")
     list(JOIN MESON_CFLAGS_${_config} ", " MESON_CFLAGS_${_config})
     string(REPLACE "'', " "" MESON_CFLAGS_${_config} "${MESON_CFLAGS_${_config}}")
     string(APPEND NATIVE_${_config} "c_args = [${MESON_CFLAGS_${_config}}]\n")
-    string(REGEX REPLACE "( |^)(-|/)" ";\\2" MESON_CXXFLAGS_${_config} "${VCPKG_DETECTED_CMAKE_CXX_FLAGS_${_config}}")
+    string(REGEX REPLACE "( |^)(--?|/)" ";\\2" MESON_CXXFLAGS_${_config} "${VCPKG_DETECTED_CMAKE_CXX_FLAGS_${_config}}")
```
Comment (Neumann-A) [inline scripts/cmake/vcpkg_configure_meson.cmake:302]:
```suggestion
    string(REGEX REPLACE "( +|^)(--?|/)" ";\\2" MESON_CXXFLAGS_${_config} "${VCPKG_DETECTED_CMAKE_CXX_FLAGS_${_config}}")
```

File: scripts/cmake/vcpkg_configure_meson.cmake [hunk 6]
```diff
@@ -101,15 +102,15 @@ function(vcpkg_internal_meson_generate_native_file_config _config) #https://meso
...
+    string(REGEX REPLACE "( |^)(--?|/)" ";\\2" MESON_CFLAGS_${_config} "${VCPKG_DETECTED_CMAKE_C_FLAGS_${_config}}")
     list(TRANSFORM MESON_CFLAGS_${_config} APPEND "'")
     list(TRANSFORM MESON_CFLAGS_${_config} PREPEND "'")
     #list(APPEND MESON_CFLAGS_${_config} "${LIBPATH_${_config}}")
     list(APPEND MESON_CFLAGS_${_config} "'-I${CURRENT_INSTALLED_DIR}/include'")
     list(JOIN MESON_CFLAGS_${_config} ", " MESON_CFLAGS_${_config})
     string(REPLACE "'', " "" MESON_CFLAGS_${_config} "${MESON_CFLAGS_${_config}}")
     string(APPEND NATIVE_${_config} "c_args = [${MESON_CFLAGS_${_config}}]\n")
-    string(REGEX REPLACE "( |^)(-|/)" ";\\2" MESON_CXXFLAGS_${_config} "${VCPKG_DETECTED_CMAKE_CXX_FLAGS_${_config}}")
+    string(REGEX REPLACE "( |^)(--?|/)" ";\\2" MESON_CXXFLAGS_${_config} "${VCPKG_DETECTED_CMAKE_CXX_FLAGS_${_config}}")
```
Comment (Neumann-A) [inline scripts/cmake/vcpkg_configure_meson.cmake:113]:
```suggestion
    string(REGEX REPLACE "( +|^)(--?|/)" ";\\2" MESON_CXXFLAGS_${_config} "${VCPKG_DETECTED_CMAKE_CXX_FLAGS_${_config}}")
```

File: scripts/cmake/vcpkg_configure_meson.cmake [hunk 7]
```diff
@@ -101,15 +102,15 @@ function(vcpkg_internal_meson_generate_native_file_config _config) #https://meso
     set(LIBPATH_${_config} "'${L_FLAG}${CURRENT_INSTALLED_DIR}${PATH_SUFFIX_${_config}}/lib'")
     
     set(NATIVE_${_config} "[properties]\n") #https://mesonbuild.com/Builtin-options.html
-    string(REGEX REPLACE "( |^)(-|/)" ";\\2" MESON_CFLAGS_${_config} "${VCPKG_DETECTED_CMAKE_C_FLAGS_${_config}}")
+    string(REGEX REPLACE "( |^)(--?|/)" ";\\2" MESON_CFLAGS_${_config} "${VCPKG_DETECTED_CMAKE_C_FLAGS_${_config}}")
```
Comment (Neumann-A) [inline scripts/cmake/vcpkg_configure_meson.cmake:105]:
```suggestion
    string(REGEX REPLACE "( +|^)(--?|/)" ";\\2" MESON_CFLAGS_${_config} "${VCPKG_DETECTED_CMAKE_C_FLAGS_${_config}}")
```

File: scripts/cmake/vcpkg_configure_meson.cmake [hunk 8]
```diff
@@ -36,6 +36,7 @@ This command supplies many common arguments to Meson. To see the full list, exam
 * [libepoxy](https://github.com/Microsoft/vcpkg/blob/master/ports/libepoxy/portfile.cmake)
 #]===]
 
+
```
Comment (Neumann-A) [inline scripts/cmake/vcpkg_configure_meson.cmake:39]:
```suggestion
```

File: scripts/cmake/vcpkg_install_meson.cmake [hunk 9]
```diff
@@ -81,4 +102,5 @@ function(vcpkg_install_meson)
             unset(ENV{MACOSX_DEPLOYMENT_TARGET})
         endif()
     endif()
+
```
Comment (Neumann-A) [inline scripts/cmake/vcpkg_install_meson.cmake:105]:
```suggestion
```

File: scripts/cmake/vcpkg_configure_meson.cmake [hunk 10]
```diff
@@ -273,7 +273,7 @@ function(vcpkg_internal_meson_generate_cross_file _additional_binaries) #https:/
     string(APPEND CROSS "cpu_family = '${BUILD_CPU_FAM}'\n")
     string(APPEND CROSS "cpu = '${BUILD_CPU}'\n")
     
-    if(NOT BUILD_CPU_FAM STREQUAL HOST_CPU_FAM)
+    if(NOT BUILD_CPU_FAM MATCHES "${HOST_CPU_FAM}" OR VCPKG_TARGET_IS_UWP)
```
Comment (ras0219-msft) [inline scripts/cmake/vcpkg_configure_meson.cmake:276]: Why is this `MATCHES` over `STREQUAL`?

File: scripts/cmake/vcpkg_configure_meson.cmake [hunk 11]
```diff
@@ -395,14 +382,30 @@ function(vcpkg_configure_meson)
...
+    if(NOT VCPKG_MESON_NATIVE_FILE AND NOT VCPKG_MESON_CROSS_FILE)
+        vcpkg_internal_meson_generate_native_file("_vcm_ADDITIONAL_NATIVE_BINARIES")
+    endif()
+    if(NOT VCPKG_MESON_NATIVE_FILE_DEBUG AND NOT VCPKG_MESON_CROSS_FILE)
+        vcpkg_internal_meson_generate_native_file_config(DEBUG)
+    endif()
+    if(NOT VCPKG_MESON_NATIVE_FILE_RELEASE AND NOT VCPKG_MESON_CROSS_FILE)
+        vcpkg_internal_meson_generate_native_file_config(RELEASE)
+    endif()
+    if(VCPKG_MESON_NATIVE_FILE AND NOT VCPKG_MESON_CROSS_FILE)
```
Comment (ras0219-msft) [inline scripts/cmake/vcpkg_configure_meson.cmake:402]: If these are both defined, why do we pass `none.txt`?

File: scripts/cmake/vcpkg_configure_meson.cmake (hunk 10 above)
Comment (Neumann-A) [reply scripts/cmake/vcpkg_configure_meson.cmake:276]: So that `x64_x86` MATCHES `x86` and becomes a native build. Same for `arm64` MATCHES `arm`.
UWP on the other hand should always be a crossbuild.

File: scripts/cmake/vcpkg_configure_meson.cmake (hunk 11 above)
Comment (Neumann-A) [reply scripts/cmake/vcpkg_configure_meson.cmake:402]: I assumed that vcpkg does not want to allow mixing of native&cross builds. Either the build is native or it is a cross build depending on the used triplet. 
I also did not test if meson behaves sane in install here meaning if it is only installing cross libraries or mixing cross and native binaries.

File: scripts/cmake/vcpkg_install_meson.cmake [hunk 12]
```diff
@@ -58,7 +75,7 @@ function(vcpkg_install_meson)
             string(REGEX REPLACE ".a$" ".lib" LIBNAMENEW "${LIBNAME}")
             string(REGEX REPLACE "^lib" "" LIBNAMENEW "${LIBNAMENEW}")
             file(RENAME "${_library}" "${LIBDIR}/${LIBNAMENEW}")
-            # For cmake fixes.
+            # For cmake fixes.
```
Comment (Neumann-A) [inline scripts/cmake/vcpkg_install_meson.cmake:78]:
```suggestion
            # For cmake fixes.
```

File: scripts/cmake/vcpkg_configure_meson.cmake (hunk 10 above)
Comment (ras0219-msft) [reply scripts/cmake/vcpkg_configure_meson.cmake:276]: This will likely need a follow-up PR to handle the more generic BUILD_OS != HOST_OS (such as windows cross to linux or vice versa).

What are the consequences of making _everything_ a cross build?

**Review by ras0219-msft — APPROVED** [review]
Looks good so far; still needs merge from master due to @post-due's fixes

File: scripts/cmake/vcpkg_configure_meson.cmake (hunk 10 above)
Comment (Neumann-A) [reply scripts/cmake/vcpkg_configure_meson.cmake:288]: > What are the consequences of making everything a cross build?

That probably strongly depends on whatever is in the meson.build. But I would assume that cross builds will, in general, not try to build host/native tools/codegenerator which is a problem if you start to depend on them being available within vcpkg. If everything is a cross build and the buildscript just tests for 'cross' for building the required tool than additional patches are necessary to introduce the required options/fixes to make the port actually build the tools if it is native build. Furthermore there might be more introspection into the system if it is a native build. 

But I agree everything should eventually be treated as a crossbuild. A native build is just a special case of a cross build, where the build machines happens to be the same as the target machine. Unfortunately most developers don't thing in that way and treat the cross build as the special case.

**Comment by Neumann-A** [comment]
@ras0219 Sry, I think you have to re-review. I changed quite a bit in the merge since I was sick of the amount of duplication in the flags code. There is still quite a bit of duplication and I think that maybe the native and cross generators could be merged a bit more in a future PR. 
The main problem is that the meson docs are very insufficient in what belongs where and most of the stuff is just me try&erroring it.

**Comment by JackBoosY** [comment]
Wait for https://github.com/microsoft/vcpkg/pull/16270 merge.

**Comment by JackBoosY** [comment]
Due to my lack of meson knowledge, I prefer @ras0219 @ras0219-msft to review this PR again.

**Comment by JackBoosY** [comment]
Ping @ras0219 @ras0219-msft for review this PR again.

File: scripts/cmake/vcpkg_configure_meson.cmake [hunk 13]
```diff
@@ -92,51 +95,55 @@ function(vcpkg_internal_meson_generate_native_file _additional_binaries) #https:
...
     set(${_out_var} "${${_out_var}}" PARENT_SCOPE)
 endfunction()
 
-function(vcpkg_internal_meson_generate_native_file_config _config) #https://mesonbuild.com/Native-environments.html
+function(vcpkg_internal_meson_convert_list_to_python_array _out_var)
+    set(FLAG_LIST ${ARGN})
+    list(TRANSFORM FLAG_LIST APPEND "'")
+    list(TRANSFORM FLAG_LIST PREPEND "'")
+    list(JOIN FLAG_LIST ", " ${_out_var})
+    string(REPLACE "'', " "" ${_out_var} "${${_out_var}}") # remove empty elements if any
```
Comment (ras0219-msft) [inline scripts/cmake/vcpkg_configure_meson.cmake:110]: This won't remove a single empty element.

It might be easier to construct this list using `string(APPEND)` and a foreach loop.

File: scripts/cmake/vcpkg_configure_meson.cmake [hunk 14]
```diff
@@ -278,54 +285,18 @@ function(vcpkg_internal_meson_generate_cross_file _additional_binaries) #https:/
     string(APPEND CROSS "cpu_family = '${BUILD_CPU_FAM}'\n")
     string(APPEND CROSS "cpu = '${BUILD_CPU}'\n")
 
-    if(NOT BUILD_CPU_FAM STREQUAL HOST_CPU_FAM OR VCPKG_TARGET_IS_ANDROID OR VCPKG_TARGET_IS_IOS)
+    if(NOT BUILD_CPU_FAM MATCHES "${HOST_CPU_FAM}" OR VCPKG_TARGET_IS_ANDROID OR VCPKG_TARGET_IS_IOS)
```
Comment (ras0219-msft) [inline scripts/cmake/vcpkg_configure_meson.cmake:288]: Why the change to MATCHES? It seems like `STREQUAL` would be better (don't want arm matching arm64)

File: scripts/cmake/vcpkg_configure_meson.cmake [hunk 15]
```diff
@@ -92,51 +95,55 @@ function(vcpkg_internal_meson_generate_native_file _additional_binaries) #https:
...
+    set(FLAG_LIST ${ARGN})
+    list(TRANSFORM FLAG_LIST APPEND "'")
+    list(TRANSFORM FLAG_LIST PREPEND "'")
+    list(JOIN FLAG_LIST ", " ${_out_var})
+    string(REPLACE "'', " "" ${_out_var} "${${_out_var}}") # remove empty elements if any
+    set(${_out_var} "[${${_out_var}}]" PARENT_SCOPE)
+endfunction()
+
+# Generates the required compiler properties for meson
+macro(vcpkg_internal_meson_generate_flags_properties_string _out_var _config)
```
Comment (ras0219-msft) [inline scripts/cmake/vcpkg_configure_meson.cmake:115]:
```suggestion
function(vcpkg_internal_meson_generate_flags_properties_string _out_var _config)
```

File: scripts/cmake/vcpkg_configure_meson.cmake [hunk 16]
```diff
@@ -92,51 +95,55 @@ function(vcpkg_internal_meson_generate_native_file _additional_binaries) #https:
...
-    list(TRANSFORM LINKER_FLAGS_${_config} PREPEND "'")
     list(APPEND LINKER_FLAGS_${_config} "${LIBPATH_${_config}}")
-    list(JOIN LINKER_FLAGS_${_config} ", " LINKER_FLAGS_${_config})
-    string(REPLACE "'', " "" LINKER_FLAGS_${_config} "${LINKER_FLAGS_${_config}}")
-    string(APPEND NATIVE_${_config} "c_link_args = [${LINKER_FLAGS_${_config}}]\n")
-    string(APPEND NATIVE_${_config} "cpp_link_args = [${LINKER_FLAGS_${_config}}]\n")
+    vcpkg_internal_meson_convert_list_to_python_array(LINKER_FLAGS_${_config} ${LINKER_FLAGS_${_config}})
+    string(APPEND ${_out_var} "c_link_args = ${LINKER_FLAGS_${_config}}\n")
+    string(APPEND ${_out_var} "cpp_link_args = ${LINKER_FLAGS_${_config}}\n")
+endmacro()
```
Comment (ras0219-msft) [inline scripts/cmake/vcpkg_configure_meson.cmake:141]:
```suggestion
    set(${_out_var} "${${_out_var}}" PARENT_SCOPE)
endfunction()
```

**Review by ras0219-msft — APPROVED** [review]
This generally LGTM. I'll wait to merge until after you've replied to my comments.

File: scripts/cmake/vcpkg_configure_meson.cmake (hunk 13 above)
Comment (Neumann-A) [reply scripts/cmake/vcpkg_configure_meson.cmake:110]: > This won't remove a single empty element.

It will remove empty strings `""`. It should be noted here that the elements are already whitespace striped because otherwise meson will pass them wrongly to the tool invocation. 

> It might be easier to construct this list using string(APPEND) and a foreach loop.

Yeah I also thought about that but did not want to deal with strange corner cases if there are elements containing `;`. So I thought `list(TRANSFORM)` is a saver approach

File: scripts/cmake/vcpkg_configure_meson.cmake (hunk 14 above)
Comment (Neumann-A) [reply scripts/cmake/vcpkg_configure_meson.cmake:288]: It is the other way around.
BUILD = arm64 | x64_x86
HOST = arm | x86 
arm64 matches arm but not the other way around
the same for x64_x86 which matches x86 but not the other way.

Comment (Neumann-A) [reply scripts/cmake/vcpkg_configure_meson.cmake:289]: Also if `VCPKG_CROSSCOMPILING` is a thing now I could use it here.

File: scripts/cmake/vcpkg_configure_meson.cmake [hunk 17]
```diff
@@ -278,54 +286,18 @@ function(vcpkg_internal_meson_generate_cross_file _additional_binaries) #https:/
     string(APPEND CROSS "cpu_family = '${BUILD_CPU_FAM}'\n")
     string(APPEND CROSS "cpu = '${BUILD_CPU}'\n")
 
-    if(NOT BUILD_CPU_FAM STREQUAL HOST_CPU_FAM OR VCPKG_TARGET_IS_ANDROID OR VCPKG_TARGET_IS_IOS)
+    if(NOT BUILD_CPU_FAM MATCHES "${HOST_CPU_FAM}" OR VCPKG_TARGET_IS_ANDROID OR VCPKG_TARGET_IS_IOS)
```
Comment (Neumann-A) [inline scripts/cmake/vcpkg_configure_meson.cmake:289]:
```suggestion
    if(NOT BUILD_CPU_FAM MATCHES "${HOST_CPU_FAM}" OR VCPKG_TARGET_IS_ANDROID OR VCPKG_TARGET_IS_IOS OR VCPKG_TARGET_IS_UWP)
```

**Review by vicroms — APPROVED** [review]
LGTM

**Review by JackBoosY — APPROVED** [review]
